### PR TITLE
feat: add files sharing restriction - audio files SQSERVICES-1555

### PR DESCRIPTION
**Pull request by MarcinR**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1555" title="SQSERVICES-1555" target="_blank">TaskSQSERVICES-1555</a>  BSI ::: iOS column 1 (non-federation) build with appropriate file sharing restrictions
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Created mechanism for restricting some kinds of  files for flag from API and for Security flag (FILE_SHARING_ENABLED) and implemented it for audio files.

### CODE REVIEW

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+class MediaShareRestrictionManager {
+    private let sessionRestriction: SessionFileRestrictionsProtocol?
+    
+    init(sessionRestriction: SessionFileRestrictionsProtocol?) {
+        self.sessionRestriction = sessionRestriction
+    }
+    
+    private let securityFlagRestrictedTypes: [ShareableMediaSource] = [.photoLibrary, .shareExtension, .clipboard]
+    
+    var mediaShareRestrictionLevel: MediaShareRestrictionLevel {
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:47]: There are several Swift lint issues

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+class MediaShareRestrictionManager {
+    private let sessionRestriction: SessionFileRestrictionsProtocol?
+    
+    init(sessionRestriction: SessionFileRestrictionsProtocol?) {
+        self.sessionRestriction = sessionRestriction
+    }
+    
+    private let securityFlagRestrictedTypes: [ShareableMediaSource] = [.photoLibrary, .shareExtension, .clipboard]
+
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:46]: There are several Swift lint issues

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    
+    init(sessionRestriction: SessionFileRestrictionsProtocol?) {
+        self.sessionRestriction = sessionRestriction
+    }
+    
+    private let securityFlagRestrictedTypes: [ShareableMediaSource] = [.photoLibrary, .shareExtension, .clipboard]
+    
+    var mediaShareRestrictionLevel: MediaShareRestrictionLevel {
+        guard let sessionRestriction = sessionRestriction, sessionRestriction.sharingEnabled() == false else {
+            return .APIFlag
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:49]: Should be `return .none` because sharingEnabled() is true in this case

Comment (MarcinR) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:49]: but "sessionRestriction.sharingEnabled() == true" mean that functionality (sharing files) is enabled, so there are no. API restrictions.
Shortening all the rest is good idea.

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift (hunk 2 above)
Comment (MarcinR) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:46]: I don't see any. Can you copy it to me (or send screenshot)

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift (hunk 3 above)
Comment (MarcinR) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:49]: fixed

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift (hunk 2 above)
Comment (MarcinR) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift]: done

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    case camera
+    case photoLibrary
+    case sketch
+    case gif
+    case audioRecording
+    case shareExtension
+    case clipboard
+}
+
+enum MediaShareRestrictionLevel {
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:32]: @johnxnguyen What do you think of case titles?

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    case none
+    case securityFlag
+    case APIFlag
+}
+
+class MediaShareRestrictionManager {
+    private let sessionRestriction: SessionFileRestrictionsProtocol?
+    private let securityFlagRestrictedTypes: [ShareableMediaSource] = [.photoLibrary, .shareExtension, .clipboard]
+
+    init(sessionRestriction: SessionFileRestrictionsProtocol?) {
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:42]: @johnxnguyen This has been implemented for userSession and sharedSession.

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift (hunk 4 above)
Comment (johnxnguyen) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift:32]: Which titles? `securityFlag` and `APIFlag`?

File: Wire-iOS/Sources/Helpers/SessionFileRestrictionsProtocol.swift [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireSyncEngine
+import WireShareEngine
+
+protocol SessionFileRestrictionsProtocol {
+    func sharingEnabled() -> Bool
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Wire-iOS/Sources/Helpers/SessionFileRestrictionsProtocol.swift]: Done

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        self.sessionRestriction = sessionRestriction
+    }
+
+    var mediaShareRestrictionLevel: MediaShareRestrictionLevel {
+        if let sessionRestriction = sessionRestriction, sessionRestriction.sharingEnabled() == false {
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift]: Done

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        case .none:
+            return true
+        case .securityFlag:
+            return securityFlagRestrictedTypes.contains(source)
+        case .APIFlag:
+            return false
+        }
+    }
+
+    func canDownloadMedia() -> Bool {
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift]: Done

File: Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        return  SecurityFlags.fileSharing.isEnabled
+    }
+
+    func canCopyFromClipboard() -> Bool {
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Wire-iOS/Sources/Helpers/MediaShareRestrictionManager.swift]: Done

**Review by KaterinaWire — APPROVED** [review]

File: Wire-iOS.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -1239,7 +1243,6 @@
 		EE01D08825ADA3D5001DB205 /* AppLockModule.Interactor.swift in Sources */ = {isa = PBXBuildFile; fileRef = EE01D08325ADA3D5001DB205 /* AppLockModule.Interactor.swift */; };
 		EE01D08925ADA3D5001DB205 /* AppLockModule.View.swift in Sources */ = {isa = PBXBuildFile; fileRef = EE01D08425ADA3D5001DB205 /* AppLockModule.View.swift */; };
 		EE09DE6025B24B2000B515F0 /* AppLockModuleRouterTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = EE09DE5E25B24B2000B515F0 /* AppLockModuleRouterTests.swift */; };
-		EE12922827E47E6D00C9CC60 /* Wire Share Extension.appex in Embed App Extensions */ = {isa = PBXBuildFile; fileRef = 168A16A91D9597C2005CFA6C /* Wire Share Extension.appex */; settings = {ATTRIBUTES = (RemoveHeadersOnCopy, ); }; };
```
Comment (KaterinaWire) [inline Wire-iOS.xcodeproj/project.pbxproj:1242]: We should not remove this

File: Wire-iOS.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -7288,7 +7295,6 @@
 				EE373FA923FBF438008C1E52 /* ConversationInputBarViewControllerDelegateTests.swift */,
 				EFC57A0F21BABCC90008DBE6 /* ConversationInputBarViewControllerAudioRecorderSnapshotTests.swift */,
 				061275D626F23535006E8D4C /* ConversationInputBarViewController+DropInteractionTests.swift */,
-				70C5560C27B2E6B100119A7D /* MockClassificationProvider.swift */,
```
Comment (KaterinaWire) [inline Wire-iOS.xcodeproj/project.pbxproj:7291]: We should not remove this

**Review by johnxnguyen — APPROVED** [review]

File: Wire-iOS.xcodeproj/project.pbxproj (hunk 10 above)
Comment (MarcinR) [reply Wire-iOS.xcodeproj/project.pbxproj]: fixed

**Pull request merged**